src/jdk.management/unix/classes/com/sun/management/internal/OperatingSystemImpl.java

```diff
@@ -136,12 +136,11 @@ public double getCpuLoad() {
         if (containerMetrics != null) {
             long quota = containerMetrics.getCpuQuota();
             if (quota > 0) {
-                long periodLength = containerMetrics.getCpuPeriod();
                 long numPeriods = containerMetrics.getCpuNumPeriods();
                 long usageNanos = containerMetrics.getCpuUsage();
-                if (periodLength > 0 && numPeriods > 0 && usageNanos > 0) {
-                    long elapsedNanos = TimeUnit.MICROSECONDS.toNanos(periodLength * numPeriods);
-                    double systemLoad = (double) usageNanos / elapsedNanos;
+                if (numPeriods > 0 && usageNanos > 0) {
+                    long quotaNanos = TimeUnit.MICROSECONDS.toNanos(quota * numPeriods);
```

Comment:
We happened to hit an exactly similar problem when running on a container with openjdk15.

Given we effectively agree that the problem is `elapsedNanos` doesn't accurately reflect the cpu time allocated across all shares vs a single share, my proposal was to use `getCpuShares` as a multiplier for `periodLength` above. 
Is there a good reason `getCpuQuota` is a better alternative?

Reply:
Hi Argha, thanks a lot for your suggestion. I think both "quota" and "share" are worth considering. Let us look into the implementation of `CgroupSubsystem::active_processor_count()` in OpenJDK HotSpot (https://github.com/openjdk/jdk/blob/master/src/hotspot/os/linux/cgroupSubsystem_linux.cpp).

Reply:
Thanks for linking that. It sounds reasonable to me to prefer `quota` in that case.

Reply:
> Thanks for linking that. It sounds reasonable to me to prefer `quota` in that case.

Yes, flag `PreferContainerQuotaForCPUCount` is [true by default](https://github.com/openjdk/jdk/blob/739769c8/src/hotspot/os/linux/globals_linux.hpp#L62). Therefore, [my current implementation](https://github.com/openjdk/jdk/pull/3656/commits/b052b624c84) might be a minimal implementation.

We can also cover the case where `PreferContainerQuotaForCPUCount` is false. There are two different ways.
1. To access the value of `PreferContainerQuotaForCPUCount`, so that we can decide if we should use `quota` or (`quota` &  `share`);
2. Reuse `CgroupSubsystem::active_processor_count`. However, the function returns an integer. It is more reasonable to use a floating number.

Looking forward to your suggestion.

Reply:
> Hi Argha, thanks a lot for your suggestion. I think both "quota" and "share" are worth considering.

@tanghaoth90 My local testing suggests that your fix addresses the issue of CPU quotas set via `--cpu-quota/--cpu-period`. When using `--cpu-shares` the CPU load calculation is wrong since it will (wrongly) report host values. Lets look at them individually, fix the quota and shares case individually (i.e. when not both are set). Once that's done, quota will be preferred in the OperatingSystemMXBean impl, which is reasonable. I don't think we need to account for the shares-preferred-over-quota at this point since that changed in HotSpot in JDK 11 time-frame (JDK-8197589) and OperatingSystemMXBean has only been made container aware in JDK 14 (yes, it got backported, but still).

Reply:
> @argha-c The proposed fix is within the `quota > 0` condition. I.e. this is code only run when CPU quotas, _not_ shares are in effect. In docker/podman speach these are `--cpu-quota=...` and `--cpu-period=....` switches. So no, in this case it wouldn't make sense to use cpu shares info in a branch which looks at cpu quotas ;-)

@jerboaa : Correct. My comment was less specific to the branch, and more to highlight that a fix here needs to consider the case for both `quota` and `shares`. I see the bug report has been updated to reflect that. Cheers.

Reply:
OK. Thanks for clarifying.

Reply:
@tanghaoth90

As to your observation of 1). Yes, that's true and I'm not sure it's worth changing. We could change the javadoc.

As for 2) I've created a PR for your branch with a suggested implementation for cpu shares based accounting. It uses a similar heuristics than the getCpuLoad0() uses here:
https://github.com/openjdk/jdk/blob/master/src/jdk.management/linux/native/libmanagement_ext/UnixOperatingSystem.c#L278

That's the best I could think of. It's certainly more accurate than the current implementation.

Reply:
@jerboaa Just read your comment about my observation (due to the weird format of github pull request). Sorry for making further [change](https://github.com/openjdk/jdk/pull/3656/commits/eba3bc10b4ebd) for the `quota` branch without asking for your suggestion. For now, I still hold a different opinion about the `quota` branch. My local test shows that
```
podman run --cpu-shares=4096 <image>
podman run --cpu-quota=400000 --cpu-period=100000 <image>
```
yields the same result after the change, otherwise the second one still prints the average cpu load.